packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -3731,7 +3731,17 @@ class StatefulElement extends ComponentElement {
     assert(_state._debugLifecycleState == _StateLifecycle.created);
     try {
       _debugSetAllowIgnoredCallsToMarkNeedsBuild(true);
-      _state.initState();
+      final dynamic _debugCheckForReturnedFuture = _state.initState() as dynamic;
+      assert(() {
+        if (_debugCheckForReturnedFuture is Future) {
+          throw new FlutterError(
+            '${_state.runtimeType}.didUpdateWidget returned a Future.\n'
+            'State.didUpdateWidget must be a void method without an `async` keyword.\n'
+            'Async lifecycle methods violate the State contract.\n'
```

Comment:
we should probably make this message more stylistically similar to the similar message in setState (e.g. use " rather than backticks, also include some advice as to what to do instead).

Reply:
Done